src/test/bin/quic_gtest.cpp

```diff
@@ -227,27 +229,41 @@ struct TestLoggerT {
     }
 };
 
-TEST(ParameterValidation, ValidateApi) {
-    TestLogger Logger("QuicTestValidateApi");
-    if (TestingKernelMode) {
-        ASSERT_TRUE(DriverClient.Run(IOCTL_QUIC_RUN_VALIDATE_API));
-    } else {
-        QuicTestValidateApi();
-    }
+// Helpers to invoke a test in kernel mode through the test driver.
+
+template<class FunType>
+bool InvokeKernelTest(const std::string& Name, FunType) {
+    static_assert(std::is_invocable_v<FunType>, "Invalid parameters for test function");
+    QUIC_RUN_TEST_REQUEST Request{};
+    Name.copy(Request.FunctionName, sizeof(Request.FunctionName));
+
+    return DriverClient.Run(IOCTL_QUIC_RUN_TEST, (void*)&Request, (uint32_t)sizeof(Request));
 }
 
-TEST(ParameterValidation, ValidateRegistration) {
-    TestLogger Logger("QuicTestValidateRegistration");
-    if (TestingKernelMode) {
-        ASSERT_TRUE(DriverClient.Run(IOCTL_QUIC_RUN_VALIDATE_REGISTRATION));
-    } else {
-        QuicTestValidateRegistration();
-    }
+template<class FunType, class ParamType>
+bool InvokeKernelTest(const std::string& Name, FunType, const ParamType& Params) {
+    static_assert(std::is_invocable_v<FunType, const ParamType&>, "Invalid parameters for test function");
+    static_assert(std::is_pod_v<ParamType>, "ParamType must be POD");
+
+    // Serialize the request header and arguments
+    std::array<uint8_t, sizeof(QUIC_RUN_TEST_REQUEST) + sizeof(ParamType)> Buffer{};
+    auto& Request = *reinterpret_cast<QUIC_RUN_TEST_REQUEST*>(Buffer.data());
+    Name.copy(Request.FunctionName, sizeof(Request.FunctionName));
+    Request.ParameterSize = sizeof(ParamType);
+    std::copy_n(
+        reinterpret_cast<const uint8_t*>(&Params),
+        sizeof(ParamType), Buffer.data() + sizeof(QUIC_RUN_TEST_REQUEST));
+
+    return DriverClient.Run(IOCTL_QUIC_RUN_TEST, (void*)Buffer.data(), (uint32_t)Buffer.size());
 }
 
+#define FUNC(TestFunction) \
+    #TestFunction, TestFunction
+
 TEST(ParameterValidation, ValidateGlobalParam) {
     TestLogger Logger("QuicTestValidateGlobalParam");
     if (TestingKernelMode) {
+        ASSERT_TRUE(InvokeKernelTest(FUNC(QuicTestGlobalParam)));
         ASSERT_TRUE(DriverClient.Run(IOCTL_QUIC_RUN_VALIDATE_GLOBAL_PARAM));
```

Comment:
Did you mean to delete this? Seems like this is running the test twice, once using the old way, and once with the new way.